espnet2/enh/diffusion/score_based_diffusion.py

```diff
@@ -0,0 +1,194 @@
+                    y_mini = y[i * minibatch : (i + 1) * minibatch]
+                    sampler = sampling.get_pc_sampler(
+                        predictor_name,
+                        corrector_name,
+                        sde=sde,
+                        score_fn=self.score_fn,
+                        y=y_mini,
+                        **kwargs
+                    )
+                    sample, n = sampler()
+                    samples.append(sample)
+                    ns.append(n)
+                samples = torch.cat(samples, dim=0)
+                return samples, ns
+
+            return batched_sampling_fn
+
+    def get_ode_sampler(self, y, N=None, minibatch=None, **kwargs):
+        N = self.sde.N if N is None else N
+        sde = self.sde.copy()
+        sde.N = N
+
+        kwargs = {"eps": self.t_eps, **kwargs}
+        if minibatch is None:
+            return sampling.get_ode_sampler(
+                sde, self.score_fn, y=y, device=y.device, **kwargs
+            )
+        else:
+            M = y.shape[0]
+
+            def batched_sampling_fn():
+                samples, ns = [], []
+                for i in range(int(math.ceil(M / minibatch))):
+                    y_mini = y[i * minibatch : (i + 1) * minibatch]
+                    sampler = sampling.get_ode_sampler(
+                        sde, self.score_fn, y=y_mini, device=y_mini.device, **kwargs
+                    )
+                    sample, n = sampler()
+                    samples.append(sample)
+                    ns.append(n)
+                samples = torch.cat(samples, dim=0)
+                return sample, ns
+
+            return batched_sampling_fn
+
+    def score_fn(self, x, t, y):
+        # Concatenate y as an extra channel
+        dnn_input = torch.cat([x, y], dim=1)
+
+        # the minus is most likely unimportant here - taken from Song's repo
+        score = -self.dnn(dnn_input, t)
+        return score
+
+    def forward(
+        self,
+        feature_ref,
+        feature_mix,
+    ):
+        # feature_ref: B, T, F
+        # feature_mix: B, T, F
+        x = feature_ref.permute(0, 2, 1).unsqueeze(1)
+        y = feature_mix.permute(0, 2, 1).unsqueeze(1)
+
+        t = (
+            torch.rand(x.shape[0], device=x.device) * (self.sde.T - self.t_eps)
+            + self.t_eps
+        )
+        mean, std = self.sde.marginal_prob(x, t, y)
+        z = torch.randn_like(x)  # i.i.d. normal distributed with var=0.5
+        sigmas = std[:, None, None, None]
+        perturbed_data = mean + sigmas * z
+
+        score = self.score_fn(perturbed_data, t, y)
+        err = score * sigmas + z
+        loss = self._loss(err)
+
+        return loss
+
+    def enhance(
+        self,
+        noisy_specturm,
+        sampler_type="pc",
+        predictor="reverse_diffusion",
+        corrector="ald",
+        N=30,
+        corrector_steps=1,
+        snr=0.5,
+        timeit=False,
+        **kwargs
+    ):
```

Comment:
It is better to add a doctoring to explain each argument since this method is used frequently.

Reply:
Done.